src/utils.rs

```diff
@@ -0,0 +1,281 @@
+        let missing_results = self.missing_results.load(Ordering::Relaxed);
+        let missing_buckets = self.missing_buckets.load(Ordering::Relaxed);
+        let missing_scp = self.missing_scp.load(Ordering::Relaxed);
+
+        if missing_history > 0 {
+            error!("Missing {} history files", missing_history);
+        }
+        if missing_ledger > 0 {
+            error!("Missing {} ledger files", missing_ledger);
+        }
+        if missing_transactions > 0 {
+            error!("Missing {} transactions files", missing_transactions);
+        }
+        if missing_results > 0 {
+            error!("Missing {} results files", missing_results);
+        }
+        if missing_buckets > 0 {
+            error!("Missing {} buckets", missing_buckets);
+        }
+        if missing_scp > 0 {
+            warn!("Missing {} optional scp files", missing_scp);
+        }
+    }
+
+    pub fn has_failures(&self) -> bool {
+        self.failed_files.load(Ordering::Relaxed) > 0
+    }
+}
+
+/// Fetch and validate .well-known/stellar-history.json from store
+pub async fn fetch_well_known_history_file(store: &StorageRef) -> Result<HistoryFileState, Error> {
+    use crate::history_format::ROOT_WELL_KNOWN_PATH;
+    use tokio::io::AsyncReadExt;
+    use tracing::debug;
+
+    debug!("Fetching .well-known from path: {}", ROOT_WELL_KNOWN_PATH);
+
+    // Read the file content
+    let mut reader = store.open_reader(ROOT_WELL_KNOWN_PATH).await.map_err(|e| {
+        std::io::Error::new(
+            std::io::ErrorKind::NotFound,
+            format!("Failed to open {}: {}", ROOT_WELL_KNOWN_PATH, e),
+        )
+    })?;
+    let mut buffer = Vec::new();
+    reader
+        .read_to_end(&mut buffer)
+        .await
+        .map_err(|e| std::io::Error::new(e.kind(), format!("Reading {}: {}", ROOT_WELL_KNOWN_PATH, e)))?;
+
+    // Parse the JSON
+    let state: HistoryFileState =
+        serde_json::from_slice(&buffer).map_err(|e| crate::history_format::Error::InvalidJson {
+            path: ROOT_WELL_KNOWN_PATH.to_string(),
+            error: e.to_string(),
+        })?;
+
+    // Validate the .well-known format
+    state.validate()?;
+
+    Ok(state)
+}
+
+/// Compute checkpoint bounds from source archive and user-specified low/high values
+/// Returns (first_checkpoint, final_checkpoint)
+pub async fn compute_checkpoint_bounds(
+    source: &StorageRef,
+    low: Option<u32>,
+    high: Option<u32>,
+) -> Result<(u32, u32), Error> {
+    use crate::history_format;
+    use tracing::{info, warn};
+
+    // Fetch the .well-known file from source
+    let state = fetch_well_known_history_file(source).await?;
+    let current_ledger = state.current_ledger;
+    let highest_source_checkpoint = history_format::round_to_lower_checkpoint(current_ledger);
+
+    info!(
+        "Source archive reports current ledger: {} (checkpoint: 0x{:08x})",
+        current_ledger, highest_source_checkpoint
+    );
```

Comment:
Not super important, but when playing with the command I appreciated in the trace output that the checkpoint is shown here in both integer and hex form. There are other logs where it's only shown in one or the other, sometimes in trace logs near each other, and it could be helpful to always log them both with one in brackets like this.